core/shared/src/main/scala/zio/ZQueue.scala

```diff
@@ -103,6 +103,30 @@ trait ZQueue[-RA, -RB, +EA, +EB, -A, +B] extends Serializable { self =>
    */
   def takeUpTo(max: Int): ZIO[RB, EB, List[B]]
 
+  /**
+   * Takes between min and max number of values from the queue. If there
+   * is less tham min items available, it'll block until the items are
+   * collected.
+   */
+  final def takeBetween(min: Int, max: Int): ZIO[RB, EB, List[B]] = {
+    def loop(remaining: Int, collected: Int, acc: List[B]): ZIO[RB, EB, List[B]] =
+      takeUpTo(remaining).flatMap { data =>
+        val len          = data.length
+        val newCollected = collected + len
+        val newAcc       = acc ++ data
+
+        if (newCollected >= min)
+          ZIO.succeedNow(newAcc)
+        else
+          loop(remaining - len, newCollected, newAcc)
```

Comment:
I believe this is a busy-loop, no?

Maybe you could repeat `take` for `min` times then do `takeUpTo(max - min)`?

Reply:
Well I admit both of these implementations would suffer from performance so maybe it isn't the greatest idea to expose either of them. Not sure.

Reply:
Both are spinners, yes. I could benchmark them, but I don't think we're going to get any substantial improvements there.

Reply:
I think a busy loop is not acceptable. If you `takeUpTo` and you still need to take more, I would `take` until you get to the minimum.

Or if you want to get fancy, you could `take` the first one, and then `takeBetween` the ones after that (recursively), such that it would at least _try_ to `takeUpTo` for remaining elements after something was added to the queue.

Reply:
Re-implemented as `takeUpTo + (min - size) * take`.